config/webpack/plugin/GeneratePlugin.js

```diff
@@ -0,0 +1,119 @@
+/* eslint-disable no-param-reassign */
+const HtmlWebpackPlugin = require('html-webpack-plugin');
+const path = require('path');
+
+const generateAppConfig = require('../../../scripts/generate-app-config/generateAppConfig');
+const getNewRelicJS = require('../../../scripts/new-relic/getNewRelicJS');
+
+/**
+ * Generate the html file
+ * @param {*} object
+ * filename - name for the generated html file
+ * lang - language for the page
+ * siteConfig - config for the page
+ * siteEntries - all entries
+ * entry - this entry
+ */
+const addHtmlPlugin = ({
+  filename, siteConfig, siteEntries, entry,
+}) => {
+  const otherSiteEntries = siteEntries.filter(indexEntry => indexEntry !== entry);
+  return new HtmlWebpackPlugin({
+    title: siteConfig.appConfig.title,
+    filename,
+    template: path.join(__dirname, '..', '..', '..', 'lib', 'index.html'),
+    rootElementId: 'root',
+    favicon: siteConfig.appConfig.favicon,
+    headHtml: [getNewRelicJS()].concat(siteConfig.appConfig.headHtml),
+    headChunks: ['rewriteHistory'],
+    excludeChunks: ['redirect', ...otherSiteEntries],
+    inject: false, // This turns off auto injection. We handle this ourselves in the template.
+  });
+};
+
+/**
+ * Generate the prefixed entry name
+ * @param {*} site
+ */
+const prefixEntry = site => (site.prefix ? `${site.prefix}/${'index'}` : 'index');
+
+/**
+ * Generate app title based on the app config.
+ * @param {*} site
+ */
+const appTitle = (site) => {
+  const { headline, title, subline } = site.siteConfig.appConfig;
+
+  return [headline, title, subline].filter(item => item).join(' - ');
+};
+
+/**
+ * Generate the config files needed for the sites passed in.
+ */
+class GeneratePlugin {
+  constructor({
+    sites, basename = '',
+  } = {}) {
+    this.entries = [];
+    this.apps = [];
+    this.sites = sites.map((site) => {
+      const entry = prefixEntry(site);
+      this.entries.push(entry);
+
+      let filename = 'index.html';
+      let siteBasename = basename;
+
+      // update filename and basename with the prefix for the site
+      if (site.prefix) {
+        filename = `${site.prefix}/index.html`;
+        siteBasename = `${basename}/${site.prefix}`;
+      }
+
+      // list of all apps to switch to
+      this.apps.push({
+        path: site.prefix,
+        url: siteBasename || '/',
+        title: appTitle(site),
+      });
+      return ({
+        ...site,
+        // name of webpack entry to be added
+        entry,
+        // name of html file to be generated
+        filename,
+        // basename for react router
+        basename: siteBasename,
+      });
+    });
+  }
+
+  apply(compiler) {
+    this.sites.forEach((site) => {
+      const {
+        siteConfig, prefix, basename, filename, entry, indexPath,
```

Comment:
Should we be validating each value is defined before applying the configuration ? It'd likely be more helpful to consumers to thrown an error indicating what is missing that having webpack fluke out or have the dev-site generate unexpectedly and require some time to debug

Reply:
I see some of this is done in the constructor, should the indexPath be verified or use path.resolve()?

Reply:
I'll add some verification for the site options, but I'll let webpack handle the file exists check.